### ISSUE TITLE
Emulate mouse on touch screens

### ISSUE THREAD

**Issue by nvaccessAuto** [nvaccess/nvda#3886]
**Reported by k_kolev1985 on 2014-02-14 13:53**
At present, on a touch screen, NVDA can interact with things on the screen only via object navigation. If an object is well designed, it is OK. But if not - problems arise. For example, if an object has no default action assigned, there is no way (at least as far as I know) to interact with it, only via the touch screen - NVDA will say "no action" and nothing will happen. For other cases in witch object exploring fails when working only with a touch screen, see tickets #3779 and #3780. To read text on a touch screen with gestures, there is the so called "text mode", To switch between those 2 modes (text and objects), the user must tap with 3 fingers on the touch screen.

So, to improve the touch screen experience with NVDA, I suggest that a third mode of screen exploration is added to NVDA. The mode will emulate the mouse cursor. When this new mode is activated and the user explores the screen with his finger, NVDA will behave as if the mouse cursor is following the user's finger and NVDA will read things on the screen as if the user is moving the mouse cursor on the screen with "mouse tracking" enabled. Gestures for performing various actions with the mouse buttons must be added, but I think that at least some of them (eg. left single click, right single click, mouse buttons lock) are already available and only gestures for those actions must be assigned. Also, gestures for left double click, middle button single click and wheel scrolling must also be added.

I don't know however if my suggestion of mouse emulation is even possible to implement, but I wont know until I ask - don't I (?) :).

**Comment by surfer0627** [comment]
Hi,
If there is something, caused this issue does not add to the official release?

**Comment by josephsl** [comment]
Hi,
Not yet - for the most part, it is complete - will need to test if it can be incubated safely, as it's been almost two years since commiting to that branch).
Thanks.

**Comment by josephsl** [comment]
Hi, I've sort of forgotten about this, as there are more 
important work being done at the moment.  Thanks for reminding me 
about this.  Thanks.

 ----- Original Message -----

**Comment by josephsl** [comment]
Hi,

Found it - knew there was an issue about mouse support.

At least for right mouse click, I do have a solution, which will be done as a pull request this weekend (basically transferring parts of Enhanced Touch Gestures add-on to Core).

Thanks.

**Comment by josephsl** [comment]
Hi,

Removed t3886 branch from NV Access repo so I can rebase this branch from my side on Python 3 master (correcting linting and Python 3 issues as we go).

Thanks.

**Comment by Adriani90** [comment]
While #10065 solves the mouse right click issue, there are also other mouse actions which still need to be considered and addressed. I am reopening this discussion.

**Comment by josephsl** [comment]
Hi, I advise filing separate issues. Closing (again).

**Comment by josephsl** [comment]
Hi,

Let's discuss this on separate issues for now.

**Comment by ChrisMedley** [comment]
We now have no right click touch gesture what so ever now in 2020.1 beta 1
Worked fine in 2019.3
Why has this been removed?

**Comment by josephsl** [comment]
Hi, are you using Enhanced Touch Gestures add-on by any chance? If so, try restarting NVDA with all add-ons disabled and see if that makes a difference. If it does make a difference, I’ll take a look at it. Thanks.

**Comment by ChrisMedley** [comment]
Yes I am using enhanced  touch gestures 20.03 and even with all addons disabled the right click gesture still no longer working

**Comment by josephsl** [comment]
Hi, that’s odd, as it is working on my computer. Can you try enabling input help mode (NVDA+1, or if using Enhanced Touch Gestures, four finger double tap) and see what “tap and hold with one finger” says? Or better yet, can you attach a debug log where you try right mouse click with touchscreens (restart NVDA with debug logging enabled, try a right-mouse click while NVDA is running, press NvDA+F1 to open log viewer, and copy and paste everything into a file and attach it)? Thanks.

**Comment by ChrisMedley** [comment]
OK, this is odd, I did as you asked and turned on NVDA help (nvda+1)
I tried the the single finger double tap and hold and got nothing, I then turned off nvda help and the single finger double tap and hold gesture  worked!

Again turned on nvda help and it also reported  right click this time also
And its still working after turning off nvda help 

Dont ask me what happened but  it's working now for whatever reason

apologies and thank you for your help

**Comment by josephsl** [comment]
Hi, I was about to suggest restarting your computer, but at least it is working. Let us know if it doesn’t work again. Thanks.
